react/week3/todoapp-w3/src/components/TodoList.js

```diff
@@ -0,0 +1,72 @@
+import React, { useState } from "react";
+import TodoForm from "./TodoForm";
+import Todo from "./Todo";
+import FetchAPI from "./Fetch";
+import TimeUsed from "./TimeUsed";
+import Deadline from "./DatePicker";
+import PropTypes from "prop-types";
+function TodoList() {
+  const [todos, setTodos] = useState([]);
+
+  const addTodo = (todo) => {
+    if (!todo.text || /^\s*$/.test(todo.text)) {
+      return;
+    }
+
+    const newTodos = [todo, ...todos];
+
+    setTodos(newTodos);
+    console.log(...todos);
+  };
+
+  const updateTodo = (todoId, newValue) => {
+    if (!newValue.text || /^\s*$/.test(newValue.text)) {
+      return;
+    }
+
+    setTodos((prev) =>
+      prev.map((item) => (item.id === todoId ? newValue : item))
+    );
+  };
+
+  const removeTodo = (id) => {
+    const removedArr = [...todos].filter((todo) => todo.id !== id);
+
+    setTodos(removedArr);
+  };
+
+  const completeTodo = (id) => {
+    let updatedTodos = todos.map((todo) => {
+      if (todo.id === id) {
+        todo.isComplete = !todo.isComplete;
+      }
```

Comment:
The problem is here that even though you are creating a new array with the map() function, you are still modifying the original object with the assignment operator. 

An immutable way of doing this is using the spread operator here:
```
if(todo.id === id) {
  return {...todo, isComplete: !isComplete}
} else {
  return todo
}
```

So with the spread operator first I copy the todo with the 3 dots, and after the comma I specify the key whose value I want to update.

Reply:
Thanks for the comments. I have changed the required code

Reply:
Great, it looks good. :+1: You can merge it to master.